doc/reference/trd-digest.md

```diff
@@ -0,0 +1,168 @@
+The digest will be computed only over the active range, rather than the
+entire slice.
+
+```rust
+pub trait DigestData<'a, const L: usize> {
+    fn set_data_client(&'a self, client: &'a dyn ClientData<'a, L>) {}
+    fn add_data(&self, data: LeasableBuffer<'static, u8>) 
+       -> Result<(), (ErrorCode, LeasableBuffer<'static, u8>)>;
+    fn add_mut_data(&self, data: LeasableMutableBuffer<'static, u8>)
+       -> Result<(), (ErrorCode, LeasableMutableBuffer<'static, u8>)>;
+    fn clear_data(&self);
+}
+```
+
+A successful call to `add_data` or `add_mut_data` will add all of the
+data in the active range of the leasable buffer as input to the hash
+function. A successful call is one which returns `Ok(())` and whose
+completion event passes `Ok(())`. If a client needs to compute a hash over several non-contiguous
+regions of a slice, or multiple slices, it can call these methods multiple
+times. 
+
+There may only be one outstanding `add_data`  or `add_mut_data` operation at
+any time. If either `add_data` or `add_mut_data` returns `Ok(())`, then all
+subsequent calls to `add_data` or `add_mut_data` MUST return `Err((ErrorCode::BUSY, ...))`
+until a completion callback delivered through `ClientData`.
+
+```rust
+pub trait ClientData<'a, const L: usize> {
+    fn add_data_done(&'a self, result: Result<(), ErrorCode>, data: LeasableBuffer<'static, u8>);
+    fn add_mut_data_done(
+        &'a self,
+        result: Result<(), ErrorCode>,
+        data: LeasableMutableBuffer<'static, u8>,
+    );
+}
+```
+
+The `data` parameters of `add_data_done` and `add_mut_data_done` indicate what
+data was added and what remains to be added to the digest. If either callback
+has a `result` value of `Ok(())`, then the active region of `data` MUST be zero
+length and all of the data in the active region passed through the corresponding
+call MUST have been added to the digest. 
+
+A call to `DigestData::clear_data()` terminates the current digest computation and
+clears out all internal state to start a new one.
+If there is an outstanding `add_data` or `add_data_mut` when `clear_data()` is called,
+the digest engine MUST issue a corresponding callback with an `Err(ErrorCode::CANCEL)`.
+
+A digest engine MUST accept multiple calls to `add_data` and `add_mut_data`. Each 
+call appends to the data over which the digest is computed. 
+
+3 Computing and Verification: `DigestHash`, `DigestVerify`, `ClientHash`, and `ClientVerify`
+===============================
+
+Once all of the data has been added as the input to a digest, a client can either
+compute the digest or ask the digest engine to compare its computed digest with
+a known value (verify).
+
+
+
+
+4 Composite Traits and Configuration
+===============================
+
+5 Capsules
+===============================
+
+The Tock kernel provides several capsules for digests:
+
+
+6 Authors' Address
+=================================
+
+    Alistair Sinclair
```

Comment:
```suggestion
    Alistair Francis
```
Unless Alistair has a different last name I do not know about :)

Reply:
:)